src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs

```diff
@@ -114,76 +115,84 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
-                                    && x.Name != Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName
-                                    && x.Name != Constants.Documents.Indexing.Fields.ValueFieldName
-                                    && FieldUtil.GetRangeTypeFromFieldName(x.Name) == RangeType.None)
-                        .Distinct(UniqueFieldNames.Instance)
-                        .ToDictionary(x => x.Name, x => new FieldsToFetch.FieldToFetch(x.Name, null, null, x.IsStored, isDocumentId: false, isTimeSeries: false));
-                }
 
-                if (fields == null)
-                    fields = FieldsToFetch.Fields;
-                else if (FieldsToFetch.Fields != null && FieldsToFetch.Fields.Count > 0)
+                if (retrieverInput.LuceneDocument != null)
                 {
-                    foreach (var kvp in FieldsToFetch.Fields)
+
+                    if (FieldsToFetch.ExtractAllFromIndex)
                     {
-                        if (fields.ContainsKey(kvp.Key))
-                            continue;
+                        fields = retrieverInput.LuceneDocument.GetFields()
+                            .Where(x => x.Name != Constants.Documents.Indexing.Fields.DocumentIdFieldName
+                                        && x.Name != Constants.Documents.Indexing.Fields.SourceDocumentIdFieldName
+                                        && x.Name != Constants.Documents.Indexing.Fields.ReduceKeyHashFieldName
+                                        && x.Name != Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName
+                                        && x.Name != Constants.Documents.Indexing.Fields.ValueFieldName
+                                        && FieldUtil.GetRangeTypeFromFieldName(x.Name) == RangeType.None)
+                            .Distinct(UniqueFieldNames.Instance)
+                            .ToDictionary(x => x.Name, x => new FieldsToFetch.FieldToFetch(x.Name, null, null, x.IsStored, isDocumentId: false, isTimeSeries: false));
+                    }
 
-                        fields[kvp.Key] = kvp.Value;
+                    if (fields == null)
+                        fields = FieldsToFetch.Fields;
+                    else if (FieldsToFetch.Fields != null && FieldsToFetch.Fields.Count > 0)
+                    {
+                        foreach (var kvp in FieldsToFetch.Fields)
+                        {
+                            if (fields.ContainsKey(kvp.Key))
+                                continue;
+
+                            fields[kvp.Key] = kvp.Value;
+                        }
                     }
                 }
-
                 foreach (var fieldToFetch in fields.Values)
                 {
-                    if (TryExtractValueFromIndex(fieldToFetch, input, result, state))
+                    //todo maciej: make it simpler
+                    if (retrieverInput.DocumentId == string.Empty)
```

Comment:
introduce readonly field SearchEngineType in this base class. Do switch one it.

Reply:
done